lib/phoenix/endpoint/instrumenter.ex

```diff
@@ -0,0 +1,138 @@
+defmodule Phoenix.Endpoint.Instrument do
+  @doc false
+  defmacro __using__(_opts) do
+    quote unquote: false do
+      # The `otp_app` variable is set by Phoenix.Endpoint.
+      # TODO: maybe it's better to make Phoenix endpoints export __otp_app__()
+      # or something similar?
+      config        = Application.get_env(var!(otp_app), __MODULE__, [])
+      instrumenters = config[:instrumentation] || []
+
+      unless is_list(instrumenters) and Enum.all?(instrumenters, &is_atom/1) do
+        raise ":instrumentation must be a list of instrumenter modules"
+      end
+
+      # This looks like [{Inst1, :event1}, {Inst1, :event2}, {Inst2, :event1}, ...]
+      events_to_instrumenters = Phoenix.Endpoint.Instrument.events_to_instrumenters(instrumenters)
+
+      @doc """
+      Instruments the given function.
+
+      `event` is the event identifier (usually an atom) that specifies which
+      instrumenting function to call in the instrumenter modules. `runtime` is
+      metadata to be associated with the event at runtime (e.g., the query being
+      issued if the event to instrument is a DB query).
+
+      ## Examples
+
+          instrument :render_view, [view: "index.html"], fn ->
+            render conn, "index.html"
+          end
+
+      """
+      defmacro instrument(event, runtime \\ nil, fun) do
+        compile = Macro.escape(__CALLER__)
```

Comment:
We don't need the whole __CALLER__. It may in fact be too much. Literally this: https://github.com/elixir-lang/elixir/blob/master/lib/logger/lib/logger.ex#L533-L538

(we can even copy the compile time app thing).